### PR TITLE
fix: delivery recording bug

### PR DESCRIPTION
**Pull request by kupermind**
- Fixing delivery recording bug;
- Adding more tests.

### CODE REVIEW

File: contracts/BalanceTrackerFixedPriceBase.sol [hunk 1]
```diff
@@ -109,9 +109,6 @@ abstract contract BalanceTrackerFixedPriceBase {
         // Get account balance
         uint256 balance = mapRequesterBalances[requester] + initAmount;
 
-        // Get mech max delivery rate
-        uint256 maxDeliveryRate = IMech(mech).maxDeliveryRate();
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceBase.sol:113]: This is moved into the MM itself

File: contracts/MechMarketplace.sol [hunk 2]
```diff
@@ -400,8 +393,16 @@ contract MechMarketplace is IErrorsMarketplace {
         mechDelivery.responseTimeout = responseTimeout + block.timestamp;
         // Record request account
         mechDelivery.requester = msg.sender;
-        // Record deliveryRate for request
-        mechDelivery.deliveryRate = msg.value;
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:404]: This was nasty, ugh!

**Review by Sfgangloff — APPROVED** [review]

File: contracts/BalanceTrackerFixedPriceBase.sol [hunk 3]
```diff
@@ -52,8 +52,8 @@ abstract contract BalanceTrackerFixedPriceBase {
     event Withdraw(address indexed account, address indexed token, uint256 amount);
     event Drained(address indexed token, uint256 collectedFees);
 
-    // Max marketplace fee
-    uint256 public constant MAX_FEE = 10_000;
+    // Max marketplace fee factor (100%)
+    uint256 public constant MAX_FEE_FACTOR = 10_000;
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceBase.sol:56]: Better naming now

File: contracts/BalanceTrackerFixedPriceBase.sol [hunk 4]
```diff
@@ -190,18 +189,41 @@ abstract contract BalanceTrackerFixedPriceBase {
...
     }
 
-    function _withdraw(uint256 balance) internal virtual;
+    function _withdraw(address mech, uint256 balance) internal virtual;
+
+    /// @dev Processes mech payment by mech service multisig.
+    /// @param mech Mech address.
+    /// @return Mech payment.
+    /// @return Marketplace fee.
+    function processPaymentByMultisig(address mech) external returns (uint256, uint256) {
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceBase.sol:198]: This is required IMO - mech itself or multisig operating the mech are able to withdraw. Also facilitates testing a lot.

File: contracts/BalanceTrackerFixedPriceBase.sol [hunk 5]
```diff
@@ -226,10 +248,10 @@ abstract contract BalanceTrackerFixedPriceBase {
         collectedFees += marketplaceFee;
 
         // Clear balances
-        mapMechBalances[msg.sender] = 0;
+        mapMechBalances[mech] = 0;
 
         // Process withdraw
-        _withdraw(balance);
+        _withdraw(mech, mechPayment);
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceBase.sol:254]: Duh :( another stupid bug

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**